File: TaskBoard/TaskBoardEditDialog/TaskBoardEditDialog/TaskCardEditDialog.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Drawing.Imaging;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using Telerik.WinControls;
using Telerik.WinControls.Primitives;
using Telerik.WinControls.UI;
using Telerik.WinControls.UI.TaskBoard;

namespace TaskBoardEditDialog
{
    public partial class TaskCardEditDialog : Telerik.WinControls.UI.RadForm
    {
        private RadTaskCardElement taskCardToEdit; 
        private RadTaskBoard taskBoard;
        private Size imageSize = new Size(16, 16);
        private BindingList<string> teams = new BindingList<string>() { "WinForms", "WPF", "Reporting", "Blazor", "DocumentProcessing" };

        private TaskCardEditDialog()
        {
            InitializeComponent();

            this.StartPosition = FormStartPosition.CenterParent;

            this.Text = "Edit Card";
        }

        public TaskCardEditDialog(RadTaskCardElement defaultTaskCard, RadTaskBoard taskBoardControl) : this()
        { 
            this.taskCardToEdit = defaultTaskCard;
            this.taskBoard = taskBoardControl; 

            this.usersCheckedDropDownList.DataSource = this.taskBoard.Users;
            this.usersCheckedDropDownList.ValueMember = "Initials";

            this.usersCheckedDropDownList.CheckedDropDownListElement.ItemHeight = 26;
            this.usersCheckedDropDownList.CheckedDropDownListElement.AutoCompleteEditableAreaElement.AutoCompleteTextBox.CreateTextBlock += AutoCompleteTextBox_CreateTextBlock;
            this.usersCheckedDropDownList.CheckedDropDownListElement.AutoCompleteEditableAreaElement.AutoCompleteTextBox.TextBlockFormatting += AutoCompleteTextBox_TextBlockFormatting;
            this.usersCheckedDropDownList.VisualListItemFormatting += usersCheckedDropDownList_VisualListItemFormatting;

            this.tagsAutoCompleteBox.AutoCompleteDataSource = teams;
            LoadSettings(this.taskCardToEdit);
        }

        private void LoadSettings(RadTaskCardElement taskCard)
        {
            this.titleTextBox.Text = taskCard.TitleText;
            this.descriptionTextBox.Text = taskCard.DescriptionText;
            foreach (UserInfo user in taskCard.Users)
            { 
                RadCheckedListDataItem item = this.usersCheckedDropDownList.Items.FirstOrDefault(x => x.Value.Equals(user.Initials)) as RadCheckedListDataItem;
                item.Checked = true;      
            }
            foreach (RadTaskCardTagElement tag in taskCard.TagElements)
            {
                this.tagsAutoCompleteBox.Text += tag.Text + ";";
            }
        }

        private void AutoCompleteTextBox_TextBlockFormatting(object sender, TextBlockFormattingEventArgs e)
        {
            ImageTokenizedTextBlockElement imageToken = e.TextBlock as ImageTokenizedTextBlockElement;
            if (imageToken != null)
            {
                RadCheckedListDataItem dataItem = imageToken.Item.Value as RadCheckedListDataItem;
                if (dataItem != null)
                {
                    UserInfo user = dataItem.DataBoundItem as UserInfo;
                    if (user != null)
                    {
                        imageToken.Image.Image = ResizeImage(user.Avatar, imageSize); 
                    }
                }
            }
        }

        private void AutoCompleteTextBox_CreateTextBlock(object sender, CreateTextBlockEventArgs e)
        {
            if (e.TextBlock is TokenizedTextBlockElement)
            {
                e.TextBlock = new ImageTokenizedTextBlockElement();
            }
        }

        private void usersCheckedDropDownList_VisualListItemFormatting(object sender, VisualItemFormattingEventArgs args)
        {
            UserInfo user = args.VisualItem.Data.DataBoundItem as UserInfo;
            if (user != null)
            {
                RadCheckedListVisualItem visualItem = args.VisualItem as RadCheckedListVisualItem;
                visualItem.CheckBox.Text = user.FirstName + " " + user.LastName;
                visualItem.CheckBox.Image = user.Avatar;
                visualItem.CheckBox.TextImageRelation = TextImageRelation.ImageBeforeText;
            }
        }

        private void radButtonOK_Click(object sender, EventArgs e)
        {
            this.taskCardToEdit.TitleText = this.titleTextBox.Text;
            this.taskCardToEdit.DescriptionText = this.descriptionTextBox.Text;
            this.taskCardToEdit.Users.Clear();
            foreach (RadCheckedListDataItem checkedUser in this.usersCheckedDropDownList.CheckedItems)
            {
                this.taskCardToEdit.Users.Add(checkedUser.DataBoundItem as UserInfo);
            }
            this.taskCardToEdit.TagElements.Clear();
            foreach (RadTokenizedTextItem token in this.tagsAutoCompleteBox.Items)
            {
                RadTaskCardTagElement tag = new RadTaskCardTagElement();
                tag.Text = token.Text;
                this.taskCardToEdit.TagElements.Add(tag);
            }
        }

        public class ImageTokenizedTextBlockElement : TokenizedTextBlockElement
        {
            private ImagePrimitive image;

            public ImagePrimitive Image
            {
                get
                {
                    return this.image;
                }
            }

            protected override void CreateChildElements()
            {
                base.CreateChildElements();
                this.image = new ImagePrimitive();
                this.image.ImageLayout = System.Windows.Forms.ImageLayout.Zoom;
                this.image.StretchVertically = false;
                this.image.StretchHorizontally = false;
                this.image.MaxSize = new Size(0, 20);
                this.Children.Insert(0, this.image);
            }

            protected override RadTokenizedTextItem CreateTokenizedTextItem(string text, object value)
            {
                RadTokenizedTextItem item = base.CreateTokenizedTextItem(text, value);
                return item;
            }

            protected override Type ThemeEffectiveType
            {
                get
                {
                    return typeof(TokenizedTextBlockElement);
                }
            }
        }

        public static Bitmap ResizeImage(Image image, Size s)
        {
            int width = s.Width;
            int height = s.Height;
            var destRect = new Rectangle(0, 0, width, height);
            var destImage = new Bitmap(width, height);

            destImage.SetResolution(image.HorizontalResolution, image.VerticalResolution);

            using (var graphics = Graphics.FromImage(destImage))
            {
                graphics.CompositingMode = CompositingMode.SourceCopy;
                graphics.CompositingQuality = CompositingQuality.HighQuality;
                graphics.InterpolationMode = InterpolationMode.HighQualityBicubic;
                graphics.SmoothingMode = SmoothingMode.HighQuality;
                graphics.PixelOffsetMode = PixelOffsetMode.HighQuality;

                using (var wrapMode = new ImageAttributes())
                {
                    wrapMode.SetWrapMode(WrapMode.TileFlipXY);
                    graphics.DrawImage(image, destRect, 0, 0, image.Width, image.Height, GraphicsUnit.Pixel, wrapMode);
                }
            }

            return destImage;
        }
    }
}

File: TaskBoard/TaskBoardEditDialog/TaskBoardEditDialogVB/TaskCardEditDialog.vb
Imports Telerik.WinControls.UI
Imports Telerik.WinControls.Primitives
Imports System.Drawing.Drawing2D
Imports System.Drawing.Imaging
Imports Telerik.WinControls.UI.TaskBoard
Imports System.ComponentModel

Public Class TaskCardEditDialog
    Private taskCardToEdit As RadTaskCardElement
    Private taskBoard As RadTaskBoard
    Private imageSize As Size = New Size(16, 16)
    Private teams As BindingList(Of String) = New BindingList(Of String)() From {
        "WinForms",
        "WPF",
        "Reporting",
        "Blazor",
        "DocumentProcessing"
    }

    Private Sub New()
        InitializeComponent()
        Me.StartPosition = FormStartPosition.CenterParent
        Me.Text = "Edit Card"
    End Sub

    Public Sub New(ByVal defaultTaskCard As RadTaskCardElement, ByVal taskBoardControl As RadTaskBoard)
        Me.New()
        Me.taskCardToEdit = defaultTaskCard
        Me.taskBoard = taskBoardControl
        Me.usersCheckedDropDownList.DataSource = Me.taskBoard.Users
        Me.usersCheckedDropDownList.ValueMember = "Initials"
        Me.usersCheckedDropDownList.CheckedDropDownListElement.ItemHeight = 26
        AddHandler Me.usersCheckedDropDownList.CheckedDropDownListElement.AutoCompleteEditableAreaElement.AutoCompleteTextBox.CreateTextBlock, AddressOf AutoCompleteTextBox_CreateTextBlock
        AddHandler Me.usersCheckedDropDownList.CheckedDropDownListElement.AutoCompleteEditableAreaElement.AutoCompleteTextBox.TextBlockFormatting, AddressOf AutoCompleteTextBox_TextBlockFormatting
        AddHandler Me.usersCheckedDropDownList.VisualListItemFormatting, AddressOf usersCheckedDropDownList_VisualListItemFormatting
        Me.tagsAutoCompleteBox.AutoCompleteDataSource = teams
        LoadSettings(Me.taskCardToEdit)
    End Sub

    Private Sub LoadSettings(ByVal taskCard As RadTaskCardElement)
        Me.titleTextBox.Text = taskCard.TitleText
        Me.descriptionTextBox.Text = taskCard.DescriptionText

        For Each user As UserInfo In taskCard.Users
            Dim item As RadCheckedListDataItem = TryCast(Me.usersCheckedDropDownList.Items.FirstOrDefault(Function(x) x.Value.Equals(user.Initials)), RadCheckedListDataItem)
            item.Checked = True
        Next

        For Each tag As RadTaskCardTagElement In taskCard.TagElements
            Me.tagsAutoCompleteBox.Text += tag.Text & ";"
        Next
    End Sub

    Private Sub AutoCompleteTextBox_TextBlockFormatting(ByVal sender As Object, ByVal e As TextBlockFormattingEventArgs)
        Dim imageToken As ImageTokenizedTextBlockElement = TryCast(e.TextBlock, ImageTokenizedTextBlockElement)

        If imageToken IsNot Nothing Then
            Dim dataItem As RadCheckedListDataItem = TryCast(imageToken.Item.Value, RadCheckedListDataItem)

            If dataItem IsNot Nothing Then
                Dim user As UserInfo = TryCast(dataItem.DataBoundItem, UserInfo)

                If user IsNot Nothing Then
                    imageToken.Image.Image = ResizeImage(user.Avatar, imageSize)
                End If
            End If
        End If
    End Sub

    Private Sub AutoCompleteTextBox_CreateTextBlock(ByVal sender As Object, ByVal e As CreateTextBlockEventArgs)
        If TypeOf e.TextBlock Is TokenizedTextBlockElement Then
            e.TextBlock = New ImageTokenizedTextBlockElement()
        End If
    End Sub

    Private Sub usersCheckedDropDownList_VisualListItemFormatting(ByVal sender As Object, ByVal args As VisualItemFormattingEventArgs)
        Dim user As UserInfo = TryCast(args.VisualItem.Data.DataBoundItem, UserInfo)

        If user IsNot Nothing Then
            Dim visualItem As RadCheckedListVisualItem = TryCast(args.VisualItem, RadCheckedListVisualItem)
            visualItem.CheckBox.Text = user.FirstName & " " + user.LastName
            visualItem.CheckBox.Image = user.Avatar
            visualItem.CheckBox.TextImageRelation = TextImageRelation.ImageBeforeText
        End If
    End Sub

    Private Sub radButtonOK_Click(ByVal sender As Object, ByVal e As EventArgs) Handles radButtonOK.Click
        Me.taskCardToEdit.TitleText = Me.titleTextBox.Text
        Me.taskCardToEdit.DescriptionText = Me.descriptionTextBox.Text
        Me.taskCardToEdit.Users.Clear()

        For Each checkedUser As RadCheckedListDataItem In Me.usersCheckedDropDownList.CheckedItems
            Me.taskCardToEdit.Users.Add(TryCast(checkedUser.DataBoundItem, UserInfo))
        Next

        Me.taskCardToEdit.TagElements.Clear()

        For Each token As RadTokenizedTextItem In Me.tagsAutoCompleteBox.Items
            Dim tag As RadTaskCardTagElement = New RadTaskCardTagElement()
            tag.Text = token.Text
            Me.taskCardToEdit.TagElements.Add(tag)
        Next
    End Sub

    Public Class ImageTokenizedTextBlockElement
        Inherits TokenizedTextBlockElement

        Private _image As ImagePrimitive

        Public ReadOnly Property Image As ImagePrimitive
            Get
                Return Me._image
            End Get
        End Property

        Protected Overrides Sub CreateChildElements()
            MyBase.CreateChildElements()
            Me._image = New ImagePrimitive()
            Me._image.ImageLayout = System.Windows.Forms.ImageLayout.Zoom
            Me._image.StretchVertically = False
            Me._image.StretchHorizontally = False
            Me._image.MaxSize = New Size(0, 20)
            Me.Children.Insert(0, Me._image)
        End Sub

        Protected Overrides Function CreateTokenizedTextItem(ByVal text As String, ByVal value As Object) As RadTokenizedTextItem
            Dim item As RadTokenizedTextItem = MyBase.CreateTokenizedTextItem(text, value)
            Return item
        End Function

        Protected Overrides ReadOnly Property ThemeEffectiveType As Type
            Get
                Return GetType(TokenizedTextBlockElement)
            End Get
        End Property
    End Class

    Public Shared Function ResizeImage(ByVal image As Image, ByVal s As Size) As Bitmap
        Dim width As Integer = s.Width
        Dim height As Integer = s.Height
        Dim destRect = New Rectangle(0, 0, width, height)
        Dim destImage = New Bitmap(width, height)
        destImage.SetResolution(image.HorizontalResolution, image.VerticalResolution)

        Using graphics = System.Drawing.Graphics.FromImage(destImage)
            graphics.CompositingMode = CompositingMode.SourceCopy
            graphics.CompositingQuality = CompositingQuality.HighQuality
            graphics.InterpolationMode = InterpolationMode.HighQualityBicubic
            graphics.SmoothingMode = SmoothingMode.HighQuality
            graphics.PixelOffsetMode = PixelOffsetMode.HighQuality

            Using wrapMode = New ImageAttributes()
                wrapMode.SetWrapMode(System.Drawing.Drawing2D.WrapMode.TileFlipXY)
                graphics.DrawImage(image, destRect, 0, 0, image.Width, image.Height, GraphicsUnit.Pixel, wrapMode)
            End Using
        End Using

        Return destImage
    End Function
End Class
